Update documentation about how to cancel/kill invocations (#239)
restatedev/documentation: docs/services/invocation.md

## Cancel an invocation

> If an invocation takes too long to complete and, therefore, is no longer of interest, you can cancel it.

Reviewer: ```suggestion
If an invocation takes too long to complete or is no longer of interest, you can cancel it.
```
There can be many reasons for cancelling besides taking too long, I guess.

Author: True. Will change it.

Canceling an invocation allows it to free any resources it is holding and roll back any changes it has made so far.
In order to roll back changes correctly, the service handlers need to contain the necessary compensation logic.

> If the required compensation logic is implemented, then the service state stays consistent even in the presence of cancellations.

Reviewer: I could interpret this as that the service state refers to Restate's K/V state, and that I need to compensate `ctx.set`.

Reviewer: ooh okay, good to know

> If the required compensation logic is implemented, then the service state stays consistent even in the presence of cancellations.

Reviewer: Do the docs explain somewhere how to implement compensation logic? Might be good to add otherwise?

The cancellation process works recursively in the following way:

> First, Restate tries to cancel the leaves of the current invocation, i.e. complete pending sleeps and awakeables or try to cancel calls to other services.

Reviewer: ```suggestion
First, Restate tries to cancel the leaves of the current invocation, i.e. interrupt ongoing sleeps and awakeables or try to cancel calls to other services.
```
The word `complete` probably only makes sense if you know that we use the term "completion" to complete a journal entry? I think interrupt might fit better here?

Author: True. Will change it.

> Once the leaves are canceled, the service handler will be notified about the cancellation via a terminal error being thrown at the call site.

Reviewer: ```suggestion
Once the leaves are canceled, a terminal error is thrown in the service handler at the point in the code the invocation had reached.
```

This allows the handler to run its specific compensation logic.
The response of the handler will then be propagated back to its caller where it will continue with the cancellation process.

:::note

> Canceling an invocation is a non-blocking operation which in some rare cases needs to be retried by the user.

Reviewer: I don't really understand what is meant here.

Author: Cancellation is a non-blocking operation which means that it returns without the operation being complete. Sometimes cancellations have no effect which means that users have to trigger the cancellation again. If this is more helpful, then I'll update the part.

:::

```shell

> $ curl -X DELETE <RESTATE_META_ENDPOINT>/invocations/<INVOCATION_IDENTIFIER>

Reviewer: ```suggestion
$ curl -X DELETE <RESTATE_META_URL>/invocations/<INVOCATION_ID>
```

There is a PR open to unify this across the docs

Author: Good catch will update it accordingly.

```

For example:

```shell
$ curl -X DELETE http://localhost:9070/invocations/T4pIkIJIGAsBiiGDV2dxK7PkkKnWyWHE
```

For more details on the API, refer to the [admin API docs](/references/admin-api).

## Kill an invocation

In a few cases, it is not possible for Restate to cancel an invocation.
For example, if the service deployment is permanently unavailable, Restate cannot invoke the service handler to run its compensation logic which is needed to complete the cancellation.
For these cases, Restate provides the ability to kill an invocation.

> Killing an invocation means that every call in the call tree of the invocation will be stopped immediately without given the service handler a chance to react.

Reviewer: ```suggestion
Killing an invocation means that every call in the call tree of the invocation will be stopped immediately without giving the service handler a chance to react and execute compensation logic.
```

This entails that killing the invocation **will not** roll back its progress.

:::note

> Background calls and delayed calls will not be killed because they are considered detached from the originating call tree.

Reviewer: ```suggestion
One-way calls and delayed calls will not be killed because they are considered detached from the originating call tree.
```

Author: Updated.

:::

:::danger
Killing an invocation might leave the service instance in an inconsistent state, just like how killing a process in your operating system may cause the open files to become corrupted. Use it with caution and try to fix the invocation in other ways before resorting to killing it.
:::

To kill an invocation, send the following request to the Restate admin API:

```shell

> $ curl -X DELETE <RESTATE_META_ENDPOINT>/invocations/<INVOCATION_IDENTIFIER>?mode=kill

Reviewer: ```suggestion
$ curl -X DELETE <RESTATE_META_URL>/invocations/<INVOCATION_ID>?mode=kill
```

Author: Same here.
